### PR TITLE
Estimate reward costs

### PR DESCRIPTION
**Pull request by girazoki**

### CODE REVIEW

File: contracts/UsingWitnet.sol [hunk 1]
```diff
@@ -97,4 +97,14 @@ contract UsingWitnet {
   function witnetReadResult(uint256 _id) internal view returns (Witnet.Result memory) {
     return Witnet.resultFromCborBytes(wrb.readResult(_id));
   }
+
+   /**
+  * @notice Estimate the reward amount
```
Comment (clbartoli) [inline contracts/UsingWitnet.sol:102]:
```suggestion
  * @notice Estimate the reward amount.
```

File: contracts/UsingWitnet.sol [hunk 2]
```diff
@@ -97,4 +97,14 @@ contract UsingWitnet {
...
+  * @dev Call to `estimate_gas_costt` function in the WitnetRequestsBoard contract.
+  * @param _gasPrice The gas price for which we want to retrieve the estimatee.
```
Comment (clbartoli) [inline contracts/UsingWitnet.sol:104]: estimation?

File: contracts/WitnetRequestsBoard.sol [hunk 3]
```diff
@@ -582,6 +582,16 @@ contract WitnetRequestsBoard is WitnetRequestsBoardInterface {
       _request.result.length == 0;
   }
 
+  /// @dev Ensures that rewards cover at least the gas cost needed for the complete DR life cycle within the Witnet Request Board.
+  /// @param _inclusionReward The amount for rewarding the reporting of the inclusion of the data request.
+  /// @param _tallyReward The amount for rewarding the reporting of the final result (aka tally) of the data request.
+  /// @param _blockReward The amount for rewarding the reporting of the blocks containing the DR inclusion and the DR result.
+  function isPayingGasCosts(uint256 _inclusionReward, uint256 _tallyReward, uint256 _blockReward) internal {
```
Comment (clbartoli) [inline contracts/WitnetRequestsBoard.sol:589]: In caase it fails, we could include the info about estimating the rewards by calling the estimageGas function

File: contracts/UsingWitnet.sol [hunk 4]
```diff
@@ -97,4 +97,14 @@ contract UsingWitnet {
   function witnetReadResult(uint256 _id) internal view returns (Witnet.Result memory) {
     return Witnet.resultFromCborBytes(wrb.readResult(_id));
   }
+
+   /**
+  * @notice Estimate the reward amount
+  * @dev Call to `estimate_gas_costt` function in the WitnetRequestsBoard contract.
```
Comment (mariocao) [inline contracts/UsingWitnet.sol:103]:
```suggestion
  * @dev Call to `estimate_gas_cost` function in the WitnetRequestsBoard contract.
```

File: contracts/WitnetRequestsBoard.sol (hunk 3 above)
Comment (girazoki) [reply contracts/WitnetRequestsBoard.sol:589]: Included

File: contracts/UsingWitnet.sol [hunk 5]
```diff
@@ -97,4 +97,14 @@ contract UsingWitnet {
   function witnetReadResult(uint256 _id) internal view returns (Witnet.Result memory) {
     return Witnet.resultFromCborBytes(wrb.readResult(_id));
   }
+
+   /**
```
Comment (mariocao) [inline contracts/UsingWitnet.sol:101]:
```suggestion
 /**
```

File: contracts/WitnetRequestsBoard.sol [hunk 6]
```diff
@@ -463,6 +463,16 @@ contract WitnetRequestsBoard is WitnetRequestsBoardInterface {
...
   }
 
+  /// @dev Estimate the amount of reward we need to insert for a given gas price.
+  /// @param _gasPrice The gas price for which we need to calculate the rewards.
+  /// @return The rewards to be included for the given gas price as inclusionReward, resultReward, blockReward.
+  function estimateGasCost(uint256 _gasPrice) public view override returns(uint256, uint256, uint256){
+    return (SafeMath.mul(_gasPrice, MAX_CLAIM_DR_GAS + MAX_DR_INCLUSION_GAS), 
+      SafeMath.mul(_gasPrice, MAX_REPORT_RESULT_GAS),  
+      SafeMath.mul(_gasPrice, MAX_REPORT_BLOCK_GAS * 2)
+      );
```
Comment (mariocao) [inline contracts/WitnetRequestsBoard.sol:473]:
```suggestion
    return (
      SafeMath.mul(_gasPrice, MAX_CLAIM_DR_GAS + MAX_DR_INCLUSION_GAS), 
      SafeMath.mul(_gasPrice, MAX_REPORT_RESULT_GAS),  
      SafeMath.mul(_gasPrice, MAX_REPORT_BLOCK_GAS * 2)
    );
```

File: contracts/WitnetRequestsBoardProxy.sol [hunk 7]
```diff
@@ -105,6 +105,14 @@ contract WitnetRequestsBoardProxy {
     return wrbWithResult.readResult(_id - offSetWrb);
   }
 
+  /// @dev Estimate the amount of reward we need to insert for a given gas price.
+  /// @param _gasPrice The gas price for which we need to calculate the rewards.
+  /// @return The rewards to be included for the given gas price as inclusionReward, resultReward, blockReward.
+  function estimateGasCost(uint256 _gasPrice) external view returns(uint256, uint256, uint256) {
+    return witnetRequestsBoardInstance.estimateGasCost(_gasPrice);
+
```
Comment (mariocao) [inline contracts/WitnetRequestsBoardProxy.sol:113]:
```suggestion
```

File: test/using_witnet.js [hunk 8]
```diff
@@ -394,6 +394,18 @@ contract("UsingWitnet", accounts => {
         "The sum of rewards overflows"
       )
     })
+    it("Should be able to estimate gas cost and post the DR", async () => {
```
Comment (mariocao) [inline test/using_witnet.js:397]:
```suggestion

    it("Should be able to estimate gas cost and post the DR", async () => {
```

Comment (mariocao) [reply test/using_witnet.js:397]: Line 385 too

**Review by mariocao — APPROVED** [review]

**Review by mariocao — APPROVED** [review]

**Pull request merged**